README.md
# pulse-block

A block for counting pulses on a Raspberry Pi GPIO pin.

# Features
- counts pulses as fast as a few hundred per second
- provides pulses per second, pulses per minute, and pulses per hour
- provides a running pulse count that can be reset via a dedicated gpio pin
- allows for use of a multiplier value to be appled to the pulse counts
- rudimentary software-based switch debouncing
- count data published via mqtt or http
- integrates well with other balena blocks, such as Connector and Dashboard

# Usage

## docker-compose file
To use this image, create a container in your `docker-compose.yml` file as shown below:

```
version: '2'

services:
  pulse:
    # Replace <arch> with your specified architecture, armv7hf for a raspi 3, or aarch64 for a raspi 4
    image: bh.cr/balenalabs/pulse-<arch>
    restart: always
    privileged: true
    labels:
      io.balena.features.supervisor-api: '1'
    expose:
      - '7575'
```

Note that the container must be privileged in order to access the GPIO pins, as well as include the balena supervisor-api label to access the device's api features.

## Data
The counter data is available as json either as an mqtt payload or via the built-in  HTTP server. To use mqtt provide an address for the `MQTT_ADDRESS` service variable. If no mqtt address is set, the  HTTP server will be available on port 7575. To force the HTTP server to be active along with mqtt, set the `ALWAYS_USE_HTTPSERVER` service variable to `True`.

The JSON will be in the following format:
`{"uuid": "a74be18a3fdc0b3bfb510e2cf84d6008", "gpio": 37, "pulse_per_second": 3, "pulse_per_minute": 180, "pulse_per_hour": 3145, "pulse_count": 3150, "pps_mult": 3, "ppm_mult": 180, "pph_mult": 3145}` 

The values ending in `_mult` are pulse counts multiplied by the optional `PULSE_MULTIPLIER` service variable. For instance, say you are counting the pulses per minute from a geiger counter and you'd like to also measure micro-Sieverts per hour (uSv/hr). You simply multiply the pulses_per_minute by a ratio specific to your geiger counter tube. (For the popular J305 tube, that ratio is 0.00812037037037) Simply enter this value as the `PULSE_MULTIPLIER` and each time the pulse_per_second, pulse_per_minute and pulse_per_hour is measured, those values times the multiplier will also be calculated. If no multiplier is provided, it defaults to a multiplier of 1.

## Pulses, debouncing, and accuracy
The pulse counter is most accurate when counting pulses from an electronic source such as another computer or device providing a pwm (pulse width modulation) or square wave signal. When counting pulses from a mechanical contact such as a physical switch, multiple unwanted pulses are often generated. These pulses can be minimized or eliminated in hardware through a properly designed "R-C Circuit" consisting of one or more resistors and capacitors. Alternatively, debouncing can be achieved through software methods. This block includes rudimentary software debouncing that should be sufficient for mechanically-generated pulses that are less (slower) than three pulses per second (3 Hz). This is accomplished by delaying the pulse count for a specified number of milliseconds (1000th of a second) after the rising edge of a pulse is detected. A good starting value is 75 milliseconds, and then slowly increasing the value by 50 until you find one that works. (This debouncing method requires smaller values than others.) You may not find a value that avoids miscounting pulses in which case you may need to figure out why your switch is so noisy and improve it somehow. To define the debounce time, set the `BOUNCE_TIME` service variable. Setting the value to `0` or deleting it removes any debouncing (the default and recommended value for counting electronic pulses.)

The default setup for the specified pulse pin is to be "floating" which is ideal for connecting to electronic devices that provide a pulsed output. To invoke one of the internal pull resistors, set the service variable `PULL_UP_DOWN` to `UP` or `DOWN`. If you are using a purely mechanical switch, you should set this value.

_NOTE_: As with most embedded Linux devices, pulse counting may not be accurate or real-time due to the OS, hardware, software or other processes running on the device. If you need strict real-time accuracy without missing even one pulse, you should use a microcontroller instead. This block is not intended for critical pulse counting applications.

## Pulse count reset
You can cause the running pulse count, `pulse_count` to reset to zero by sending a logic high pulse to the pin set by device variable `GPIO_RESET_PIN` which has a debounce value set to 200 ms, so a pushbutton switch should suffice.

## Service/device variables

| Variable | Description | Default value |
| -------- | ----------- | ------------- |
| `GPIO_PIN` | Sets the pin to use for counting incoming pulses, using BOARD numbering | 37 |
| `PULL_UP_DOWN` | Sets the selected gpio pin's internal pull up or pull down resistor. Set to either `NONE`, `UP` or `DOWN` | `NONE` (floating)|
| `GPIO_RESET_PIN` | Sets the pin used to reset the running pulse count. Normally pulled down and debounced 200 ms | 38|
| `PULSE_MULTIPLIER` | Multiplies the pulses per second, minute, and hour by this factor as pps_mult, ppm_mult and pph_mult respectively | 1 |
| `BOUNCE_TIME` | Number of milliseconds to wait before counting a pulse, used to debounce noisy mechanical switches | 0 |
| `MQTT_ADDRESS` | Provide the address of an MQTT broker for the block to publish data to. If this variable is not set and a container on the device is named mqtt, it will publish to that instead. Either disables the internal HTTP server unless `ALWAYS_USE_HTTPSERVER` is set to `True` | none |
| `ALWAYS_USE_HTTPSERVER` | Set to True to enable the internal HTTP server even when it is automatically disabled due to the detection of mqtt | 0 |


## How it works
This block utilizes the "raspberry-gpio-python" module, commonly known as [RPi.GPIO](https://sourceforge.net/projects/raspberry-gpio-python/). Specifically, the event_detected() function is used in a while loop that counts the pulses. A separate thread runs each second to aggregate the pulses counted in that second and add them to a running queue so that the pulses per minute and pulses per hour can be calculated. When activated, a minimal HTTP server runs in its own thread using a simple socket to respond to client requests.

## Use with other blocks
The pulse block works well with our [connector block](https://github.com/balena-labs-projects/connector) and [dashboard block](https://github.com/balena-labs-projects/dashboard). Connect them together to quickly create dashboards for your pulse data.

We also have an [LED Display](https://github.com/balenalabs-incubator/led-sensor-display) project that complements this block by allowing you to view the pulse data on a four digit seven segment LED.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.0.2
## (2022-11-16)

* moving from balenablocks to balena-labs-projects [Flynn Joffray]
* patch: repo file [Phil Wilson]
* patch: add build script [Phil Wilson]


build-images.sh
#!/bin/bash
set -e
BLOCK_NAME="pulse"

function build_and_push_image () {
  local DOCKER_REPO=$1
  local BALENA_MACHINE_NAME=$2
  local DOCKER_ARCH=$3
  local BALENA_ARCH=$4

  echo "Building for machine name $BALENA_MACHINE_NAME, platform $DOCKER_ARCH, pushing to $DOCKER_REPO/$BLOCK_NAME"

  sed "s/%%BALENA_MACHINE_NAME%%/$BALENA_MACHINE_NAME/g" ./Dockerfile.template > ./Dockerfile.$BALENA_MACHINE_NAME
  sed -i.bak "s/%%BALENA_ARCH%%/$BALENA_ARCH/g" ./Dockerfile.$BALENA_MACHINE_NAME && rm ./Dockerfile.$BALENA_MACHINE_NAME.bak
  docker buildx build -t $DOCKER_REPO/$BLOCK_NAME:$BALENA_MACHINE_NAME --load --platform $DOCKER_ARCH --file Dockerfile.$BALENA_MACHINE_NAME .

  echo "Publishing..."
  docker push $DOCKER_REPO/$BLOCK_NAME:$BALENA_MACHINE_NAME

  echo "Cleaning up..."
  rm Dockerfile.$BALENA_MACHINE_NAME
}

function create_and_push_manifest() {
  docker manifest create $DOCKER_REPO/$BLOCK_NAME:latest --amend $DOCKER_REPO/$BLOCK_NAME:raspberrypi3 --amend $DOCKER_REPO/$BLOCK_NAME:raspberrypi4-64
  docker manifest push $DOCKER_REPO/$BLOCK_NAME:latest
}

# YOu can pass in a repo (such as a test docker repo) or accept the default
DOCKER_REPO=${1:-balenablocks}

build_and_push_image $DOCKER_REPO "raspberrypi4-64" "linux/arm64" "aarch64"

build_and_push_image $DOCKER_REPO "raspberrypi3" "linux/arm/v7" "armv7hf"



create_and_push_manifest

server.py
import RPi.GPIO as GPIO
import time
import datetime
import os
#from balena import Balena
import sys
import paho.mqtt.client as mqtt
import json
import threading
import signal
import socket

pulse_per_second= 0
pulse_count = 0
pulse_output = {
    "uuid": os.environ.get('BALENA_DEVICE_UUID'),
    "gpio": 0,
    "pulse_per_second": 0,
    "pulse_per_minute": 0,
    "pulse_per_hour": 0,
    "pulse_count": 0,
    "pps_mult": 0,
    "ppm_mult": 0,
    "pph_mult": 0
}
env_vars = {
    "pulse_multiplier": 1,
    "gpio_pin": 37,
    "bounce_time": 200,
    "mqtt_address": "none",
    "gpio_reset_pin": 38,
    "enable_webserver": 0,
    "pull_up_down": "down"
}
sum_queue = []
pulse_multiplier = 0
client = mqtt.Client()

# Use the sdk to get services (eventually)
def mqtt_detect():
    
    return False

# Simple webserver
def background_web(server_socket):
    global pulse_output

    while True:
        # Wait for client connections
        client_connection, client_address = server_socket.accept()

        # Get the client request
        request = client_connection.recv(1024).decode()
        print(request)

        # Send HTTP response
        response = 'HTTP/1.0 200 OK\n\n'+ json.dumps(pulse_output)
        print(pulse_output)
        client_connection.sendall(response.encode())
        client_connection.close()

class ProgramKilled(Exception):
    """
    An instance of this custom exception class will be thrown every time we get an SIGTERM or SIGINT
    """
    pass

# Raise the custom exception whenever SIGINT or SIGTERM is triggered
def signal_handler(signum, frame):
    raise ProgramKilled

# This method fires on edge detection from a reset button
def on_reset(channel):
    global pulse_count
    print("pulse reset detected")
    pulse_count = 0

# This function serves as the callback triggered on every run of our IntervalThread
def action() :
    global pulse_per_second, sum_queue, client, env_vars, pulse_output
    pulse_per_minute = 0
    pulse_per_hour = 0
    pulse_multiplier = env_vars["pulse_multiplier"]
    sum_queue.append(pulse_per_second)
    if len(sum_queue) > 1:
        pulse_per_minute = sum(sum_queue[-60:])
        pulse_per_hour = sum(sum_queue[-3600:])
    if len(sum_queue) > 3600:
        sum_queue.pop(0)
    pulse_output["gpio"] = env_vars["gpio_pin"]
    pulse_output["pulse_per_second"] = pulse_per_second
    pulse_output["pulse_per_minute"] = pulse_per_minute
    pulse_output["pulse_per_hour"] = pulse_per_hour
    pulse_output["pulse_count"] = pulse_count
    pulse_output["pps_mult"] = pulse_per_second * pulse_multiplier
    pulse_output["ppm_mult"] = pulse_per_minute * pulse_multiplier
    pulse_output["pph_mult"] = pulse_per_hour * pulse_multiplier
    #print(pulse_output)
    if env_vars["mqtt_address"] != "none":
        client.publish('pulse_data', json.dumps(pulse_output))
    pulse_per_second = 0

# See https://stackoverflow.com/questions/2697039/python-equivalent-of-setinterval
class IntervalThread(threading.Thread) :
    def __init__(self,interval,action, *args, **kwargs) :
        super(IntervalThread, self).__init__()
        self.interval=interval
        self.action=action
        self.stopEvent=threading.Event()
        self.start()

    def run(self) :
        nextTime=time.time()+self.interval
        while not self.stopEvent.wait(nextTime-time.time()) :
            nextTime+=self.interval
            self.action()

    def cancel(self) :
        self.stopEvent.set()

def main():

    global pulse_per_second, pulse_count, env_vars, client

    # device variables
    env_vars["pulse_multiplier"] = float(os.getenv('PULSE_MULTIPLIER', '1'))
    env_vars["gpio_pin"] = os.getenv('GPIO_PIN', 37)
    env_vars["bounce_time"]  = os.getenv('BOUNCE_TIME', 0)
    env_vars["mqtt_address"] = os.getenv('MQTT_ADDRESS', 'none')
    env_vars["gpio_reset_pin"] = os.getenv('GPIO_RESET_PIN', 38)
    env_vars["enable_webserver"] = os.getenv('ALWAYS_USE_HTTPSERVER', 0)
    env_vars["pull_up_down"] = os.getenv('PULL_UP_DOWN', 'NONE')

    if env_vars["enable_webserver"] == "1":
        env_vars["enable_webserver"] = "True"
    else:
        env_vars["enable_webserver"] = "False"
        
    GPIO.setmode(GPIO.BOARD)
    gpio_pin = int(env_vars["gpio_pin"])

    # Set the pin for the incoming pulse
    if env_vars["pull_up_down"] == "UP":
        GPIO.setup(gpio_pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        test_cond = GPIO.LOW
    elif env_vars["pull_up_down"] == "DOWN":
        GPIO.setup(gpio_pin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
        test_cond = GPIO.HIGH
    else:
        GPIO.setup(gpio_pin, GPIO.IN)
        test_cond = GPIO.HIGH

    if env_vars["bounce_time"] == 0:
        bounce_time = 0
    else:
        if str(env_vars["bounce_time"]).isnumeric():
            bounce_time = int(env_vars["bounce_time"])/1000
            print("Bounce time set to {0} second(s)".format(bounce_time))
        else:
            bounce_time = 0

    # Set the pin for pulse count reset
    GPIO.setup(int(env_vars["gpio_reset_pin"]), GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
    GPIO.add_event_detect(int(env_vars["gpio_reset_pin"]), GPIO.FALLING, callback=on_reset, bouncetime=200)

    if mqtt_detect() and env_vars["mqtt_address"] == "none":
        env_vars["mqtt_address"] = "mqtt"

    if env_vars["mqtt_address"] != "none":
        print("Starting mqtt client, publishing to {0}:1883".format(env_vars["mqtt_address"]))
        try:
             client.connect(env_vars["mqtt_address"], 1883, 60)
        except Exception as e:
            print("Error connecting to mqtt. ({0})".format(str(e)))
            env_vars["mqtt_address"] = "none"
            env_vars["enable_webserver"] = "True"
        else:
            client.loop_start()
    else:
        env_vars["enable_webserver"] = "True"

    if env_vars["enable_webserver"] == "True":
        SERVER_HOST = '0.0.0.0'
        SERVER_PORT = 7575

        # Create socket
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind((SERVER_HOST, SERVER_PORT))
        server_socket.listen(1)
        print("HTTP server listening on port {0}...".format(SERVER_PORT))

        t = threading.Thread(target=background_web, args=(server_socket,))
        t.start()

    # Handle SIGINT and SIFTERM with the help of the callback function
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    # start action every 1s
    inter=IntervalThread(1,action)

    # See https://www.g-loaded.eu/2016/11/24/how-to-terminate-running-python-threads-using-signals/
    while True:
        try:
            GPIO.wait_for_edge(gpio_pin, GPIO.FALLING)
            if bounce_time > 0:
                time.sleep(bounce_time)
                if GPIO.input(gpio_pin) == test_cond:
                    pulse_per_second = pulse_per_second + 1
                    pulse_count = pulse_count + 1
            else:
                pulse_per_second = pulse_per_second + 1
                pulse_count = pulse_count + 1
        except ProgramKilled:
            print("Program killed: running cleanup code")
            inter.cancel()
            break

if __name__ == "__main__":
    main()
